[AutoFill Debugging] Part 1: Refactor `-_requestTextExtraction:` so that it takes a configuration object

whsieh opened the pull request:
#### 0a2be6f329bedd4f14756472e6515ff5b929025c
<pre>
[AutoFill Debugging] Part 1: Refactor `-_requestTextExtraction:` so that it takes a configuration object
<a href="https://bugs.webkit.org/show_bug.cgi?id=296864">https://bugs.webkit.org/show_bug.cgi?id=296864</a>

Reviewed by Ryosuke Niwa and Abrar Rahman Protyasha.

This is the first part of a series of patches to provide clients with additional means to debug and
address AutoFill compatibility or performance. This will build off of `WKTextExtractionItem` and
related classes (currently only used for `UIIntelligenceSupport` framework integration), to provide
a lightweight snapshot of webpage content.

For now, we simply maintain existing behavior, but expose a new `_WKTextExtractionConfiguration`
object — as SPI for now, but hopefully eventually API — that clients can use to configure how text
extraction should behave (i.e., only by specifying a `targetRect` for now).

In the process, we also shuffle around some of the sources in `UIProcess/Cocoa/TextExtraction`, in
preparation for further changes to these files. In particular, we add `_WKTextExtraction.h` to
`API/` and move `WKTextExtractionItem.h` there as `_WKTextExtractionInternal.h`, as the internal
header counterpart to the private header. The corresponding Swift file is also moved, to keep the
relevant code in the same directory.

Instead of returning a `WKTextExtractionItem` in the completion handler, `-_requestTextExtraction:`
also now returns a `WKTextExtractionResult` object, which wraps the root item. While that&apos;s all it
currently does, we&apos;ll extend it in a subsequent patch to contain additional information, such as
native popup menu items.

* Source/WebKit/Modules/Internal/WebKitInternal.h:
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h: Copied from Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:

Add a new source file to contain the ObjC implementation of several exported `_WKTextExtraction*`
classes (for now, just the configuration). This is in a separate file rather than in Swift because
Swift sources are disabled for non-internal tvOS and watchOS builds, which would result in an error
that `_WKTextExtractionConfiguration` is declared but not defined when building for pure-OpenSource
tvOS or watchOS.

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift: Renamed from Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionItem.swift.
(_WKTextExtractionConfiguration.targetRect):
(_WKTextExtractionConfiguration.mergeParagraphs):
(_WKTextExtractionConfiguration.ignoreTransparency):

Implement `_WKTextExtractionConfiguration`. These two flags are not used yet, but will be once we
introduce support for `WebCore::TextExtraction::Request` in a subsequent change.

(WKTextExtractionEditable.isSecure):
(WKTextExtractionEditable.isFocused):
(WKTextExtractionLink.url):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h: Renamed from Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionItem.h.
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKWebView+TextExtraction.swift:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:

Add `API/Cocoa` to the header search path so that we can import _WKTextExtractionInternal.h directly
in WebKitTestRunner (rather than re-declare all of the internal interfaces).

* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::requestTextExtraction):
* Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.h:
* Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.mm:
(WTR::description):

Canonical link: <a href="https://commits.webkit.org/298218@main">https://commits.webkit.org/298218@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3d623a9b1a84d80dc1b5f249825a524b1a0f3ee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114647 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34393 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24856 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120813 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65369 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6c2d8e89-860b-4a8d-83e4-ea433e11480f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116536 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35021 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42953 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87148 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42050 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2c90ed82-311c-4b52-ad0b-35fb31dde30a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117595 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27906 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102964 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67536 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27093 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21088 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64490 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97286 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21198 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124015 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41659 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31114 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95962 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42036 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99150 "") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95743 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24385 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40910 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18755 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37758 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41537 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47049 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41123 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44430 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42873 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:
```diff
@@ -0,0 +1,44 @@
+#import <Foundation/Foundation.h>
+#import <WebKit/WKFoundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
```
we should prefer to use NS_HEADER_AUDIT_BEGIN(nullability, sendability) over this for better safety when interoperating with Swift

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
```diff
@@ -0,0 +1,43 @@
+#import "config.h"
+#import "_WKTextExtractionInternal.h"
+
+@implementation _WKTextExtractionConfiguration
```
this could trivially be implemented in Swift, but it's ok this way too

whsieh replied:
Yeah, I fell back to ObjC to keep the OpenSource-only tvOS/watchOS builds passing. I think once we're able to build swift on those platforms without the internal SDK, we can move this into the swift file.

rr-codes commented on Source/WebKit/UIProcess/Cocoa/TextExtraction/WKWebView+TextExtraction.swift:
```diff
@@ -104,8 +107,12 @@ extension WKWebView {
 
-            if let item = await _requestTextExtraction(visibleRect) {
-                collector.collect(createIntelligenceElement(item: item))
+            var configuration = _WKTextExtractionConfiguration()
```
this can be `let` (I think) (since it's a class so setting its properties doesn't really "count" as mutating it)

whsieh replied:
Good catch, fixed.

aprotyas commented on Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
```diff
@@ -341,11 +343,15 @@ - (void)paste:(id)sender;
-    [webView() _requestTextExtraction:extractionRect completionHandler:^(WKTextExtractionItem *item) {
+    RetainPtr configuration = adoptNS([_WKTextExtractionConfiguration new]);
+    [configuration setTargetRect:extractionRect];
+    [configuration setMergeParagraphs:YES];
+    [configuration setIgnoreTransparency:YES];
```
You don't want these to be configurable by `TextExtractionOptions`?

whsieh replied:
Ah, we do — but I was planning to do that in one of the following patches in the sequence, when we introduce support for `WebCore::TextExtraction::Request` (and we can therefore write layout tests for these behaviors).

whsieh committed:
[AutoFill Debugging] Part 1: Refactor `-_requestTextExtraction:` so that it takes a configuration object
https://bugs.webkit.org/show_bug.cgi?id=296864

Reviewed by Ryosuke Niwa and Abrar Rahman Protyasha.

This is the first part of a series of patches to provide clients with additional means to debug and
address AutoFill compatibility or performance. This will build off of `WKTextExtractionItem` and
related classes (currently only used for `UIIntelligenceSupport` framework integration), to provide
a lightweight snapshot of webpage content.

For now, we simply maintain existing behavior, but expose a new `_WKTextExtractionConfiguration`
object — as SPI for now, but hopefully eventually API — that clients can use to configure how text
extraction should behave (i.e., only by specifying a `targetRect` for now).

In the process, we also shuffle around some of the sources in `UIProcess/Cocoa/TextExtraction`, in
preparation for further changes to these files. In particular, we add `_WKTextExtraction.h` to
`API/` and move `WKTextExtractionItem.h` there as `_WKTextExtractionInternal.h`, as the internal
header counterpart to the private header. The corresponding Swift file is also moved, to keep the
relevant code in the same directory.

Instead of returning a `WKTextExtractionItem` in the completion handler, `-_requestTextExtraction:`
also now returns a `WKTextExtractionResult` object, which wraps the root item. While that's all it
currently does, we'll extend it in a subsequent patch to contain additional information, such as
native popup menu items.

* Source/WebKit/Modules/Internal/WebKitInternal.h:
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h: Copied from Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:

Add a new source file to contain the ObjC implementation of several exported `_WKTextExtraction*`
classes (for now, just the configuration). This is in a separate file rather than in Swift because
Swift sources are disabled for non-internal tvOS and watchOS builds, which would result in an error
that `_WKTextExtractionConfiguration` is declared but not defined when building for pure-OpenSource
tvOS or watchOS.

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift: Renamed from Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionItem.swift.
(_WKTextExtractionConfiguration.targetRect):
(_WKTextExtractionConfiguration.mergeParagraphs):
(_WKTextExtractionConfiguration.ignoreTransparency):

Implement `_WKTextExtractionConfiguration`. These two flags are not used yet, but will be once we
introduce support for `WebCore::TextExtraction::Request` in a subsequent change.

(WKTextExtractionEditable.isSecure):
(WKTextExtractionEditable.isFocused):
(WKTextExtractionLink.url):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h: Renamed from Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionItem.h.
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKWebView+TextExtraction.swift:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:

Add `API/Cocoa` to the header search path so that we can import _WKTextExtractionInternal.h directly
in WebKitTestRunner (rather than re-declare all of the internal interfaces).

* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::requestTextExtraction):
* Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.h:
* Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.mm:
(WTR::description):

Canonical link: https://commits.webkit.org/298218@main
